raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java

```diff
@@ -148,7 +150,7 @@ private Long append(int epoch, List<T> records, boolean isAtomic) {
                 }
 
                 if (batch == null) {
-                    return null;
+                    throw new IllegalStateException("Append failed because we failed to allocate memory to write the batch");
```

Comment:
I wonder if it is worthwhile introducing some custom raft exception types for these two cases. At a minimum, that would let us improve the assertions. Maybe `NotLeaderException` and `BufferAllocationException` or something like that?

Reply:
Thank you, I added several `RaftException` and added some unit tests, and replace 2 original `IllegalArgumentException` with them, I also considered replacing `NotLeaderOrFollowerException` but it's not very suitable, Please help to have a look.